rfcs/rdfa-editor-stage-1.md

```diff
@@ -0,0 +1,587 @@
+
+interface ModelRange {
+  start: ModelPosition;
+  end: ModelPosition;
+  direction: SelectionDirection;
+}
+```
+
+#### ModelSelection
+
+A selection represents the selected region of the document. While in practice, some browsers only
+support a single contiguous range, we follow the DOM specification which specifies that a selection
+_can_ contain multiple ranges. 
+
+```typescript 
+interface ModelSelection {
+  ranges: ModelRange[]
+}
+```
+
+#### Operation
+
+An operation represents an atomic change to the Model. Operations should be designed to be composable and combinable.
+
+Fundamentally, an insert operation which overwrites content can cover every possible editing usecase, including deletion (just insert nothing) and
+moving content (as a combination of a deletion and an insertion, aka 2 insertions).
+However, to allow for optimization at the lowest level, we define a generic operation interface rather than a single insert operation.
+
+An operation is always defined over a range. While in essence an operation is just a function, we define it as a stateful object
+which remembers its arguments to allow for easy chaining and modification before execution.
+
+
+```typescript
+interface Operation {
+  range: ModelRange;
+  execute();
+}
+```
+
+#### Mutator
+
+Whereas operations are the smallest unit of change to the document, the mutator interface represents
+the "core editing operations" needed to make useful changes to the document. Operations tend to be too
+atomic for this purpose.
+
+```typescript
+interface Mutator {
+  replaceRange(range: ModelRange, ...content: Node[]);
+  replaceNode(oldNode: Node, newNode: Node);
+  updateSelection(selection: ModelSelection);
+  selectRange(range: ModelRange);
+}
+```
+
+#### Command
+
+A command represents the highest-level primitive with which to build editing functionality. 
+Commands should be named after their intention. As certain commands only make sense in certain contexts,
+commands also carry with them a method to check whether their execution makes sense at the moment.
+
+Commands receive 2 argument objects: a config object containing the tools the command needs to interact with the document,
+and an args object containing user-defined arguments that the callsite has to provide.
+
+Commands can choose to interact with the document in 2 core ways, or a combination of both:
+- by composing other commands
+- by interacting with ranges and nodes and using the mutator to persist changes to the document
+
+```typescript
+interface CommandConfig {
+  executedBy: string;
+  controller: Controller;
+}
+
+interface Command<A> {
+  canExecute(config: CommandConfig, args: A): boolean;
+  execute(config: CommandConfig, args: A);
+}
+```
+
+### Plugins
+
+With extendability as one of the core pillars, plugin support should be considered at every level. The above
+architecture gives plugins the following capabillities:
+
+- defining new node types
+- defining new commands
+- executing commands
```

Comment:
For executing commands we will have to talk about namespacing, can a plugin call another plugin command? if so how will it be uniquely identified?

Reply:
yes, plugins can call other plugin commands
namespacing is indeed an issue with the current design.
I'd like to address that in a stage 2 rfc, I think for now we can get by with naming conventions